pkg/resources/mutation/cluster/provisioningcluster.go

```diff
@@ -53,10 +98,258 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
+	case admissionv1.Create, admissionv1.Update:
+		if cluster.DeletionTimestamp != nil {
+			return admission.ResponseAllowed(), nil
+		}
+		if templateName == "" {
+			err := m.secret.Delete(SecretNamespace, secretName, &metav1.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to delete the secret: %w", err)
+			}
+			// drop relevant fields if they exist in the cluster
+			cluster = DropMachineSelectorFile(MachineSelectorFileForPSA(secretName, mountPath), cluster)
+			args := GetKubeAPIServerArg(cluster)
+			if args[KubeAPIAdmissionConfigOption] == mountPath {
+				delete(args, KubeAPIAdmissionConfigOption)
+				cluster = SetKubeAPIServerArg(args, cluster)
+			}
+		} else {
+			// Now, handle the case of PSACT being set when creating or updating the cluster
+			template, err := m.psact.Get(templateName)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					return admission.ResponseBadRequest(err.Error()), nil
+				}
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to get psact: %w", err)
+			}
+			fileContent, err := psa.GenerateAdmissionConfigFile(template, cluster.Spec.KubernetesVersion)
+			if err != nil {
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to generate admission configuration from PSACT [%s]: %w", templateName, err)
+			}
+			anno := map[string]string{
+				SecretAnnotation: cluster.Name,
+			}
+			data := map[string]string{
+				Secretkey: fileContent,
+			}
+			_, err = m.createOrUpdateSecret(SecretNamespace, secretName, data, anno)
+			if err != nil {
+				return nil, fmt.Errorf("[provsioning cluster mutator] failed to create or update the secret for the admission configuration file: %w", err)
+			}
+			// set relevant fields in the cluster
+			cluster = AddMachineSelectorFile(MachineSelectorFileForPSA(secretName, mountPath), cluster)
+			args := GetKubeAPIServerArg(cluster)
+			args[KubeAPIAdmissionConfigOption] = mountPath
+			cluster = SetKubeAPIServerArg(args, cluster)
+		}
+	}
+	return admission.ResponseAllowed(), nil
+}
+
+// createOrUpdateSecret creates or updates a secret based on the provided information.
+func (m *ProvisioningClusterMutator) createOrUpdateSecret(namespace, name string, data, annotations map[string]string) (*corev1.Secret, error) {
+	var existing *corev1.Secret
+	var err error
+	if namespace == "" || name == "" {
+		return existing, fmt.Errorf("namespace and name are required")
+	}
+	existing, err = m.secret.Cache().Get(namespace, name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		StringData: data,
+		Type:       corev1.SecretTypeOpaque,
+	}
+
+	if annotations != nil {
+		secret.Annotations = annotations
+	}
+	if existing == nil {
+		return m.secret.Create(secret)
+	}
+	if !reflect.DeepEqual(existing.StringData, secret.StringData) {
```

Comment:
> StringData is a write-only field, so this comparison needs to be updated, as it will always be false if data != nil or {}. 

Based on my test, this is not true. Because we set the  `StringData` field when creating the Secret, it is accessible in the returned Secret object and can be used to compare. For testing purposes, I add a printing in the if clause and only see the message in the log when the StringData is changed (by changing the defaultPSACT assigned to the cluster).

Reply:
I don't think that logic holds true if the webhook restarts. Additionally, StringData is probably only populated for the secret until it is updated somewhere else. Can you verify this is working as intended after a restart?

Reply:
Here is what I found:  
Suppose others update the Secret outside of rancher-webhook, for example, by adding an annotation. In that case, the field `StringData` will be reset to its default empty value, later on, when rancher-webhook compares the existing one with the new one, it will detect the change and update the Secret even though the value is unchanged. 

Based on the above finding and your advice, I changed to logic to update and compare the `Data` field in the Secret.